cabot/cabotapp/models.py

```diff
@@ -1146,6 +1169,146 @@ def __unicode__(self):
+                                               'alerts will resume, even if the check is still failing.')
+    close_after_successes = models.PositiveIntegerField(null=True, default=1,
+                                                        help_text='After this many consecutive successful runs the '
+                                                                  'acknowledgement will be automatically closed. Enter '
+                                                                  '0 to disable.')
+
+    @classmethod
+    def get_acks_matching_check(cls, check, at_time=None):
+        # type: (StatusCheck, timezone.datetime) -> List[Acknowledgement]
+        """
+        :param check: check to gather acks for
+        :param at_time: only consider acks that were open at this time; leave as None to use timezone.now().
+        :returns list of Acknowledgements where ack.
+        """
+        at_time = at_time or timezone.now()
+        acks = cls.objects.filter(status_check_id=check.id)
+        return acks.exclude(created_at__gt=at_time).exclude(closed_at__lte=at_time).exclude(expire_at__lte=at_time)
+
+    @classmethod
+    def get_acks_matching_result(cls, result, at_time=None):
+        # type: (StatusCheckResult, Union[timezone.datetime, None]) -> List[Acknowledgement]
+        """
+        :param result: result to gather acks for
+        :param at_time: only consider acks that were open at this time; leave None for all currently open acks
+        :returns list of Acknowledgements where ack.matches_result(result) == True
+        """
+        return [a for a in cls.get_acks_matching_check(result.status_check, at_time) if a.matches_result(result)]
+
+    @classmethod
+    def close_succeeding_acks(cls, check, at_time=None):
+        # type: (StatusCheck, Union[timezone.datetime, None]) -> None
+        """
+        This function closes acks that have hit their close_after_successes threshold.
+        This should be called after a StatusCheck saves a new StatusCheckResult (including its tags!).
+        It technically only needs to be called after a check succeeds, but it's idempotent - you can call it whenever.
+        :param check: StatusCheck to update acks for.
+        :param at_time: Time to pass into get_acks_matching_check. Leave as None to use timezone.now().
+        """
+        for ack in cls.get_acks_matching_check(check, at_time=at_time):
+            if ack.close_after_successes:
+                # get last n results for our check, regardless of status
+                results = StatusCheckResult.objects.filter(status_check=ack.status_check)\
+                              .order_by('-time_complete', '-id').only('succeeded')[:ack.close_after_successes]
+
+                # now filter down to the results that succeeded
+                results = [r for r in results if r.succeeded]
+
+                # if we hit our threshold, this check should expire
+                if len(results) >= ack.close_after_successes:
```

Comment:
It shouldn't be in normal operation...but if there's some programming error that causes multiple results get created without running this function after each one, it could happen.